File: mpt_samples_worklfow/utils/wav2digi.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.stream.*;
import javax.sound.sampled.*;

/*
Processing steps:
1. Parse Args & Config: Read inputs, output file, start addr ($9000 default). Infer sample rate from ext (.d8=8kHz, .d15=15kHz). Validate or error.
2. Collect WAVs: Scan paths for .wav files (files or dirs), sort, list valid ones. Skip non-WAVs.
3. Init Structures: Prep lists for addresses, lengths, data buffer. Set current addr to start_addr.
4 .Process Each WAV:

4a: SoX convert to 8-bit mono raw PCM at target rate (temp file).
4b: Read raw, scale to 4-bit (0-15).
4c: Pack pairs into bytes ((high<<4)|low); pad odd with 0.
4d: Pad to 256-byte multiple with 0x88 (silence).
4e: Record addr/len, append to buffer, advance addr.


5 .Build Table: 32-byte array: bytes 0-15 = start high-bytes (up to 16 samples); 16-31 = end high-bytes.
6 .Write Output: Table + packed data to file. Print summary (samples, addrs, lens, size).
7 .Cleanup: Delete temps, handle errors.
*/

public class WavToDigiConverter {

    public static void main(String[] args) {
        List<String> nonOptions = new ArrayList<>();
        int startAddr = 0x9000;
        for (String arg : args) {
            if (arg.startsWith("--start-addr=")) {
                String hexStr = arg.substring(13);
                try {
                    startAddr = Integer.parseInt(hexStr, 16);
                } catch (NumberFormatException e) {
                    System.err.println("Invalid start address: " + hexStr);
                    System.exit(1);
                }
            } else {
                nonOptions.add(arg);
            }
        }

        if (nonOptions.size() < 2) {
            System.err.println("Usage: java WavToDigiConverter [input_paths...] output_file [--start-addr=0xXXXX]");
            System.exit(1);
        }

        List<String> inputPaths = nonOptions.subList(0, nonOptions.size() - 1);
        String outputFilePath = nonOptions.get(nonOptions.size() - 1);

        String outLower = outputFilePath.toLowerCase();
        float sampleRate;
        if (outLower.endsWith(".d8")) {
            sampleRate = 8000f;
        } else if (outLower.endsWith(".d15")) {
            sampleRate = 15000f;
        } else {
            System.err.println("Error: Output file must end with .d8 or .d15");
            System.exit(1);
            return; // Unreachable
        }

        List<String> wavFiles = getWavFilesFromInput(inputPaths);
        if (wavFiles.isEmpty()) {
            System.err.println("Error: No WAV files found in the provided input paths.");
            System.exit(1);
        }

        System.out.println("Found " + wavFiles.size() + " WAV files to process:");
        for (String wav : wavFiles) {
            System.out.println("  - " + wav);
        }

        try {
            convertToMultiDigi(wavFiles, outputFilePath, startAddr, sampleRate);
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }

    private static List<String> getWavFilesFromInput(List<String> inputPaths) {
        List<String> wavFiles = new ArrayList<>();
        for (String path : inputPaths) {
            Path p = Paths.get(path);
            try {
                if (Files.isRegularFile(p) && path.toLowerCase().endsWith(".wav")) {
                    wavFiles.add(path);
                } else if (Files.isDirectory(p)) {
                    try (Stream<Path> stream = Files.list(p)) {
                        List<String> dirWavs = stream
                                .filter(file -> Files.isRegularFile(file) && file.toString().toLowerCase().endsWith(".wav"))
                                .map(Path::toString)
                                .sorted()
                                .collect(Collectors.toList());
                        if (dirWavs.isEmpty()) {
                            System.out.println("Warning: No WAV files found in directory " + path);
                        } else {
                            wavFiles.addAll(dirWavs);
                        }
                    }
                } else {
                    System.out.println("Error: " + path + " is neither a file nor a directory.");
                }
            } catch (IOException e) {
                System.err.println("Error accessing path " + path + ": " + e.getMessage());
            }
        }
        return wavFiles;
    }

    private static void convertToMultiDigi(List<String> inputWavs, String outputFile, int startAddr, float sampleRate) throws IOException, UnsupportedAudioFileException {
        ByteArrayOutputStream allData = new ByteArrayOutputStream();
        List<Integer> sampleAddresses = new ArrayList<>();
        List<Integer> sampleLengths = new ArrayList<>();

        int currentAddr = startAddr;

        for (String inputWav : inputWavs) {
            File wavFile = new File(inputWav);
            AudioInputStream sourceAIS = AudioSystem.getAudioInputStream(wavFile);

            AudioFormat targetFormat = new AudioFormat(
                    AudioFormat.Encoding.PCM_UNSIGNED,
                    sampleRate,
                    8,
                    1,
                    1,
                    sampleRate,
                    false
            );

            AudioInputStream ais = AudioSystem.getAudioInputStream(targetFormat, sourceAIS);

            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            byte[] buffer = new byte[4096];
            int len;
            while ((len = ais.read(buffer)) != -1) {
                baos.write(buffer, 0, len);
            }
            ais.close();
            sourceAIS.close();
            byte[] pcmData = baos.toByteArray();

            int[] pcm4bit = new int[pcmData.length];
            for (int i = 0; i < pcmData.length; i++) {
                int sample = pcmData[i] & 0xFF;
                pcm4bit[i] = (sample / 16) & 0x0F;
            }

            int originalLength = pcm4bit.length;
            if (originalLength % 2 != 0) {
                pcm4bit = Arrays.copyOf(pcm4bit, originalLength + 1);
                pcm4bit[originalLength] = 0;
            }

            byte[] packed = new byte[pcm4bit.length / 2];
            for (int i = 0; i < packed.length; i++) {
                int s1 = pcm4bit[i * 2];
                int s2 = pcm4bit[i * 2 + 1];
                packed[i] = (byte) ((s1 << 4) | s2);
            }

            int packedLen = packed.length;
            int padding = (256 - (packedLen % 256)) % 256;
            byte[] paddedPacked = Arrays.copyOf(packed, packedLen + padding);
            for (int j = packedLen; j < paddedPacked.length; j++) {
                paddedPacked[j] = (byte) 0x88;
            }
            int paddedPackedLength = paddedPacked.length;

            sampleAddresses.add(currentAddr);
            sampleLengths.add(paddedPackedLength);

            allData.write(paddedPacked);

            currentAddr += paddedPackedLength;

            System.out.printf("Processed %s: %d 4-bit samples (packed to %d bytes, padded by %d bytes)%n",
                    inputWav, originalLength, paddedPackedLength, padding);
        }

        byte[] table = createSampleTable(sampleAddresses, sampleLengths);

        try (FileOutputStream fos = new FileOutputStream(outputFile)) {
            fos.write(table);
            allData.writeTo(fos);
        }

        System.out.println("Created " + outputFile + " with " + inputWavs.size() + " samples.");
        System.out.print("Sample addresses: ");
        for (int addr : sampleAddresses) {
            System.out.print(String.format("$%04x ", addr));
        }
        System.out.println();
        System.out.print("Sample lengths: ");
        for (int length : sampleLengths) {
            System.out.print(length + " bytes ");
        }
        System.out.println();
        System.out.println("Total file size: " + (32 + allData.size()) + " bytes");
    }

    private static byte[] createSampleTable(List<Integer> sampleAddresses, List<Integer> sampleLengths) {
        byte[] table = new byte[32];
        int num = Math.min(16, sampleAddresses.size());
        for (int i = 0; i < num; i++) {
            int addr = sampleAddresses.get(i);
            table[i] = (byte) (addr >> 8);
            int endAddr = addr + sampleLengths.get(i);
            table[16 + i] = (byte) (endAddr >> 8);
        }
        return table;
    }
}

File: mpt_samples_worklfow/utils/wav2digi.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;

/*
Processing steps:
1. Parse Args & Config: Read inputs, output file, start addr ($9000 default). Infer sample rate from ext (.d8=8kHz, .d15=15kHz). Validate or error.
2. Collect WAVs: Scan paths for .wav files (files or dirs), sort, list valid ones. Skip non-WAVs.
3. Init Structures: Prep lists for addresses, lengths, data buffer. Set current addr to start_addr.
4 .Process Each WAV:

4a: SoX convert to 8-bit mono raw PCM at target rate (temp file).
4b: Read raw, scale to 4-bit (0-15).
4c: Pack pairs into bytes ((high<<4)|low); pad odd with 0.
4d: Pad to 256-byte multiple with 0x88 (silence).
4e: Record addr/len, append to buffer, advance addr.


5 .Build Table: 32-byte array: bytes 0-15 = start high-bytes (up to 16 samples); 16-31 = end high-bytes.
6 .Write Output: Table + packed data to file. Print summary (samples, addrs, lens, size).
7 .Cleanup: Delete temps, handle errors.
*/

class Program
{
    static void Main(string[] args)
    {
        if (args.Length < 2)
        {
            Console.WriteLine("Usage: program.exe <input_paths...> <output_file> [--start-addr=0x9000]");
            return;
        }

        List<string> inputPaths = new List<string>();
        string outputFile = null;
        int startAddr = 0x9000;

        for (int i = 0; i < args.Length; i++)
        {
            if (args[i].StartsWith("--start-addr="))
            {
                string hexStr = args[i].Substring("--start-addr=".Length);
                startAddr = Convert.ToInt32(hexStr, 16);
            }
            else if (outputFile == null && i == args.Length - 1)
            {
                outputFile = args[i];
            }
            else
            {
                inputPaths.Add(args[i]);
            }
        }

        if (outputFile == null)
        {
            Console.WriteLine("Error: Output file not specified.");
            return;
        }

        int sampleRate;
        string ext = Path.GetExtension(outputFile).ToLower();
        if (ext == ".d8")
        {
            sampleRate = 8000;
        }
        else if (ext == ".d15")
        {
            sampleRate = 15000;
        }
        else
        {
            Console.WriteLine("Error: Output file must end with .d8 or .d15");
            return;
        }

        List<string> wavFiles = GetWavFilesFromInput(inputPaths);
        if (wavFiles == null || wavFiles.Count == 0)
        {
            Console.WriteLine("Error: No WAV files found in the provided input paths.");
            return;
        }

        Console.WriteLine($"Found {wavFiles.Count} WAV files to process:");
        foreach (var wav in wavFiles)
        {
            Console.WriteLine($"  - {wav}");
        }

        try
        {
            ConvertToMultiDigi(wavFiles, outputFile, startAddr, sampleRate);
        }
        catch (Exception e)
        {
            Console.WriteLine($"Error: {e.Message}");
        }
    }

    static List<string> GetWavFilesFromInput(List<string> inputPaths)
    {
        List<string> wavFiles = new List<string>();

        foreach (var path in inputPaths)
        {
            if (File.Exists(path))
            {
                if (Path.GetExtension(path).ToLower() == ".wav")
                {
                    wavFiles.Add(path);
                }
                else
                {
                    Console.WriteLine($"Warning: {path} is not a WAV file, skipping.");
                }
            }
            else if (Directory.Exists(path))
            {
                string[] dirWavs = Directory.GetFiles(path, "*.wav", SearchOption.TopDirectoryOnly);
                if (dirWavs.Length == 0)
                {
                    Console.WriteLine($"Warning: No WAV files found in directory {path}");
                }
                else
                {
                    wavFiles.AddRange(dirWavs.OrderBy(f => f));
                }
            }
            else
            {
                Console.WriteLine($"Error: {path} is neither a file nor a directory.");
                return null;
            }
        }

        return wavFiles;
    }

    static void ConvertToMultiDigi(List<string> inputWavs, string outputFile, int startAddr, int sampleRate)
    {
        List<byte> allData = new List<byte>();
        List<int> sampleAddresses = new List<int>();
        List<int> sampleLengths = new List<int>();

        int currentAddr = startAddr;

        foreach (var inputWav in inputWavs)
        {
            string tempRaw = RunSoxCommand(inputWav, sampleRate);

            try
            {
                byte[] pcmData = File.ReadAllBytes(tempRaw);

                List<byte> pcm4bit = ConvertTo4Bit(pcmData);

                if (pcm4bit.Count % 2 != 0)
                {
                    pcm4bit.Add(0);
                }

                List<byte> packedData = new List<byte>();
                for (int i = 0; i < pcm4bit.Count; i += 2)
                {
                    byte sample1 = pcm4bit[i];
                    byte sample2 = pcm4bit[i + 1];
                    byte packedByte = (byte)((sample1 << 4) | sample2);
                    packedData.Add(packedByte);
                }

                int packedLength = packedData.Count;
                int padding = (256 - (packedLength % 256)) % 256;
                packedData.AddRange(Enumerable.Repeat((byte)0x88, padding));
                int paddedPackedLength = packedData.Count;

                sampleAddresses.Add(currentAddr);
                sampleLengths.Add(paddedPackedLength);

                allData.AddRange(packedData);

                currentAddr += paddedPackedLength;

                Console.WriteLine($"Processed {inputWav}: {pcm4bit.Count} 4-bit samples " +
                                  $"(packed to {paddedPackedLength} bytes, padded by {padding} bytes)");
            }
            finally
            {
                if (File.Exists(tempRaw))
                {
                    File.Delete(tempRaw);
                }
            }
        }

        byte[] table = CreateSampleTable(sampleAddresses, sampleLengths);

        using (FileStream fs = new FileStream(outputFile, FileMode.Create))
        {
            fs.Write(table, 0, table.Length);
            fs.Write(allData.ToArray(), 0, allData.Count);
        }

        Console.WriteLine($"Created {outputFile} with {inputWavs.Count} samples.");
        Console.WriteLine($"Sample addresses: [{string.Join(", ", sampleAddresses.Select(a => $"${a:X4}"))}]");
        Console.WriteLine($"Sample lengths: [{string.Join(", ", sampleLengths.Select(l => $"{l} bytes"))}]");
        Console.WriteLine($"Total file size: {32 + allData.Count} bytes");
    }

    static string RunSoxCommand(string inputFile, int sampleRate)
    {
        string tempRaw = Path.GetTempFileName() + ".raw";

        ProcessStartInfo psi = new ProcessStartInfo
        {
            FileName = "sox",
            Arguments = $"\"{inputFile}\" -e unsigned-integer -b 8 -c 1 -r {sampleRate} \"{tempRaw}\"",
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true
        };

        using (Process process = Process.Start(psi))
        {
            string error = process.StandardError.ReadToEnd();
            process.WaitForExit();

            if (process.ExitCode != 0)
            {
                throw new Exception($"SoX error for {inputFile}: {error}");
            }
        }

        return tempRaw;
    }

    static List<byte> ConvertTo4Bit(byte[] pcmData)
    {
        List<byte> pcm4bit = new List<byte>();
        foreach (byte sample in pcmData)
        {
            byte val = (byte)((sample / 16) & 0x0F);
            pcm4bit.Add(val);
        }
        return pcm4bit;
    }

    static byte[] CreateSampleTable(List<int> sampleAddresses, List<int> sampleLengths)
    {
        byte[] table = new byte[32];

        for (int i = 0; i < Math.Min(16, sampleAddresses.Count); i++)
        {
            table[i] = (byte)(sampleAddresses[i] >> 8); // High byte of start address
        }

        for (int i = 0; i < Math.Min(16, sampleAddresses.Count); i++)
        {
            int endAddr = sampleAddresses[i] + sampleLengths[i];
            table[16 + i] = (byte)(endAddr >> 8); // High byte of end address
        }

        return table;
    }
}
